teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -129,14 +129,55 @@ Task-level Timeout (`TaskRunSpec.Timeout` and `RunSpec.Timeout`) is set for each
 
 The `PipelineTask.Retries` value, which is specified at `Pipeline` authoring time, is passed to the `TaskRunSpec.Retries` and `CustomRunSpec.Retries` during the execution of a `PipelineRun`.
 
 The `TaskRun` and `CustomRun` controllers handle their own `Retries`. 
 
 The `PipelineRun` controller does not check for `len(retriesStatus)` to determine whether a `TaskRun` or `CustomRun` is done executing. Instead, uses `ConditionSucceeded` as the only way to decide if the `TaskRun` or `CustomRun` has completed execution.
 
+#### Execution Status of a `failed` `pipelineTask` with `retries`
+
+The following table shows how an overall status of a `taskRun` or a `customRun` for a `pipelineTask` with `retries` set to 3:
+
+| `status` | `reason`             | `completionTime` is set | description                                                                                                |
+|----------|----------------------|-------------------------|------------------------------------------------------------------------------------------------------------|
+| Unknown  | Running              | No                      | The `taskRun` has been validated and started to perform its work.                                          |
+| Unknown  | CompletedWithRetries | Yes                     | The `taskRun` (zero attempt of a `pipelineTask`) finished executing but failed. It has 3 retries pending.  |
+| Unknown  | Running              | No                      | First attempt of a `taskRun` has started to perform its work.                                              |
+| Unknown  | CompletedWithRetries | Yes                     | The `taskRun` (first attempt of a `pipelineTask`) finished executing but failed. It has 2 retries pending. |
+| Unknown  | Running              | No                      | Second attempt of a `taskRun` has started to perform its work.                                             |
+| Unknown  | CompletedWithRetries | Yes                     | The `taskRun` (second attempt of a `pipelineTask`) finished executing but failed. It has 1 retry pending.  |
+| Unknown  | Running              | No                      | Third attempt of a `taskRun` has started to perform its work.                                              |
+| False    | Failed               | Yes                     | The `taskRun` (third attempt of a `pipelineTask`) finished executing but failed. No more retries pending.  |
+
+Pipeline controller can now rely on `ConditionSucceeded` set to `Failed` after all the retries are exhausted.
+
+`TaskRun` reconciler can rely on the `reason` changing from `running` to `Completed` to archive existing status into
+`retriesStatus` and clear the existing status of a `taskRun` such that a next attempt starts executing.
+
+The status of each attempt in `retriesStatus` will be set to `Failed` for clarity and backward compatibility.
+
+#### Execution Status of a `pipelineTask` with `timeout` and `retries`
```

Comment:
If I understand correctly, the execution status of a `pipelineTask` is the representation of the `TaskRun` status included in the `PipelineRun` status. 

I think this should be decided depending on the setting of the `embedded-status` feature flag.
In case of `full` it will be the same as in the `TaskRun`, so I think the table below is redundant.

Reply:
The following table represents state transition for a `pipelineTask` or an independent task when it hits the specified timeout compared to the table above which documents state transition for a failed task.

I am open to drop this table for now. This was a result of multiple discussions on intermediate status of each attempt, whether to set it to `TaskRunTimeout` or the state w finally agreed on `ToBeRetried`.